File: data_pipeline/stripe_extractor.py
# Import necessary libraries
import stripe
from google.cloud import bigquery
import os

# Set up the Stripe API key
# In a real-world scenario, you would use a secret management system to store your API key.
# For this example, we'll read it from an environment variable.
stripe.api_key = os.environ.get("STRIPE_API_KEY")

# Set up the BigQuery client
# This assumes you have authenticated with Google Cloud CLI.
client = bigquery.Client()

# Define the BigQuery table names
# In a real-world scenario, you would likely have a separate configuration file for these.
TRANSACTIONS_TABLE = "your_project.your_dataset.fct_transactions"
CUSTOMERS_TABLE = "your_project.your_dataset.dim_customers"
PRODUCTS_TABLE = "your_project.your_dataset.dim_products"

def fetch_stripe_data(endpoint, **kwargs):
    """
    Fetches data from the Stripe API using auto-pagination.
    """
    print(f"Fetching data from {endpoint}...")
    try:
        # It's better to use the specific list methods, e.g., stripe.Charge.list()
        # but for this generic example, we'll use this approach.
        data = getattr(stripe, endpoint).list(**kwargs)
        return data.auto_paging_iter()
    except Exception as e:
        print(f"Error fetching data from {endpoint}: {e}")
        return []

def load_data_to_bigquery(table_id, data):
    """
    Loads data into a BigQuery table.
    """
    print(f"Loading data into {table_id}...")
    try:
        errors = client.insert_rows_json(table_id, data)
        if not errors:
            print(f"Successfully loaded data into {table_id}")
        else:
            print(f"Errors loading data into {table_id}: {errors}")
    except Exception as e:
        print(f"Error loading data into {table_id}: {e}")


def main():
    """
    Main function to orchestrate the data extraction and loading process.
    """
    # Fetch and load balance transactions
    transactions = fetch_stripe_data("BalanceTransaction")
    # A simple transformation to match a hypothetical schema.
    # In a real pipeline, this would be more robust.
    transformed_transactions = [
        {
            "transaction_id": t.id,
            "amount": t.amount,
            "fee": t.fee,
            "net": t.net,
            "created_at": t.created,
        }
        for t in transactions
    ]
    if transformed_transactions:
        load_data_to_bigquery(TRANSACTIONS_TABLE, transformed_transactions)

    # Fetch and load customers
    customers = fetch_stripe_data("Customer")
    transformed_customers = [
        {
            "customer_id": c.id,
            "email": c.email,
            "name": c.name,
            "created_at": c.created,
        }
        for c in customers
    ]
    if transformed_customers:
        load_data_to_bigquery(CUSTOMERS_TABLE, transformed_customers)

    # Fetch and load products
    products = fetch_stripe_data("Product")
    transformed_products = [
        {
            "product_id": p.id,
            "name": p.name,
            "description": p.description,
            "created_at": p.created,
        }
        for p in products
    ]
    if transformed_products:
        load_data_to_bigquery(PRODUCTS_TABLE, transformed_products)


if __name__ == "__main__":
    main()

File: data_pipeline/stripe_extractor.js

// Import necessary libraries
const Stripe = require('stripe');
const { BigQuery } = require('@google-cloud/bigquery');

// Set up the Stripe API key
const stripe = new Stripe(process.env.STRIPE_API_KEY);

// Set up the BigQuery client
const bigquery = new BigQuery({projectId: 'ai-finance-mvp'});

// Define the BigQuery table names
const TRANSACTIONS_TABLE = 'ai-finance-mvp.ai_finance_db.fct_transactions';
const CUSTOMERS_TABLE = 'ai-finance-mvp.ai_finance_db.dim_customers';
const PRODUCTS_TABLE = 'ai-finance-mvp.ai_finance_db.dim_products';

/**
 * Fetches data from a Stripe resource using auto-pagination.
 * @param {object} resource - The Stripe resource to fetch data from (e.g., stripe.customers).
 * @returns {Promise<Array<object>>} - A promise that resolves to an array of objects.
 */
async function fetchStripeData(resource) {
  console.log(`Fetching data from Stripe for ${resource.constructor.name}...`);
  const data = [];
  try {
    for await (const item of resource.list({ limit: 100 })) {
      data.push(item);
    }
    console.log(`Successfully fetched ${data.length} items.`);
    return data;
  } catch (e) {
    console.error(`Error fetching data from Stripe: ${e.message}`);
    return [];
  }
}

/**
 * Loads data into a BigQuery table.
 * @param {string} tableId - The full ID of the table (e.g., 'your_project.your_dataset.your_table').
 * @param {Array<object>} data - The data to load.
 */
async function loadDataToBigQuery(tableId, data) {
  if (!data || data.length === 0) {
    console.log('No data to load.');
    return;
  }
  console.log(`Loading ${data.length} rows into ${tableId}...`);
  try {
    const [projectId, datasetId,tableName] = tableId.split('.');
    await bigquery.dataset(datasetId, {projectId: projectId}).table(tableName).insert(data);
    console.log(`Successfully loaded data into ${tableId}`);
  } catch (e) {
    console.error(`Error loading data into ${tableId}:`, e.response ? e.response.insertErrors : e);
  }
}

/**
 * Main function to orchestrate the data extraction and loading process.
 */
async function main() {
  // Fetch and load balance transactions
  const transactions = await fetchStripeData(stripe.balanceTransactions);
  const transformedTransactions = transactions.map(t => ({
    id: t.id,
    amount: t.amount / 100,
    created: t.created,
    currency: t.currency,
    description: t.description,
    type: t.type,
    fee: t.fee / 100,
    status: t.status
  }));
  await loadDataToBigQuery(TRANSACTIONS_TABLE, transformedTransactions);

  // Fetch and load customers
  const customers = await fetchStripeData(stripe.customers);
  const transformedCustomers = customers.map(c => ({
    id: c.id,
    name: c.name,
    email: c.email,
    created: c.created,
  }));
  await loadDataToBigQuery(CUSTOMERS_TABLE, transformedCustomers);

  // Fetch and load products
  const products = await fetchStripeData(stripe.products);
  const transformedProducts = products.map(p => ({
    id: p.id,
    name: p.name,
    description: p.description,
    created: p.created,
  }));
  await loadDataToBigQuery(PRODUCTS_TABLE, transformedProducts);
}

main().catch(console.error);
